### PR TITLE
Add SSL encrypted connections to Docker containers

### PR DESCRIPTION
**Pull request by ossarga**
- Added a common-services.yml file and reworked docker-compose.yml file so that the services are extensions of the services in common-services.yml file
- Added Generate SSL Stores service to docker-compose to generate signed certs and SSL stores
- Added Cassandra SSL service to docker-compose which only accepts encrypted JMX and Native Protocol connections
- Added Initialize Reaper DB SSL service to docker-compose to initialise keyspace in Cassandra SSL container via an encrypted connection
- Added nodetool SSL service to docker-compose to query Cassandra SSL container via an encrypted connection
- Added CQLSH SSL encrypted service to docker-compose
- Replaced intialize-reaper_db and initialize-reaper_db-ssl services to use CQLSH common service
- Added an environment variable to the Reaper Docker container to allow Reaper to use an SSL encrypted connection when connecting with Cassandra storage
- Minor updates to Docker Reaper configuration YAML so that it contains latest Reaper options
- Fixed bug where JVM options were passed to Reaper in the wrong position
- Updated site documentation to include instructions on orchestrating an SSL encrypted environment

### CODE REVIEW

File: src/packaging/docker-services/nodetool-ssl/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
+FROM cassandra-ssl
```
Comment (joaquincasares) [inline src/packaging/docker-services/nodetool-ssl/Dockerfile]: I'm unsure if `nodetool` and `nodetool-ssl` can be combined, but give it a quick go?

File: src/packaging/docker-services/reaper-ssl/reaper-ssl.env [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+REAPER_JMX_PASSWORD=reaperPass
+
+# define the Dockerized Cassandra node to serve as the Reaper backend
+# while using the reaper_db keyspace
+REAPER_STORAGE_TYPE=cassandra
+REAPER_CASS_CLUSTER_NAME="reaper-cluster"
+REAPER_CASS_CONTACT_POINTS=["cassandra-ssl"]
+REAPER_CASS_KEYSPACE=reaper_db
+
+# use SSL encryption when connecting to Cassandra via the native protocol
```
Comment (joaquincasares) [inline src/packaging/docker-services/reaper-ssl/reaper-ssl.env]: Only include what's different between `reaper` and `reaper-ssl`.

File: src/packaging/resource/ca_cert.conf [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+default_bits           = 2048
+
+[ req_distinguished_name ]
+C                      = AU
+ST                     = NSW
+L                      = Sydney
+O                      = TLP
+OU                     = TestCluster
+CN                     = TestClusterMasterCA
+emailAddress           = info@thelastpickle.com
```
Comment (joaquincasares) [inline src/packaging/resource/ca_cert.conf]: `no-reply@`?

File: src/server/pom.xml [hunk 4]
```diff
@@ -16,6 +16,7 @@
         <dropwizard.metrics.datadog.version>1.0.6</dropwizard.metrics.datadog.version>
         <cassandra.version>2.2.7</cassandra.version>
         <cucumber.version>1.2.5</cucumber.version>
+        <docker.directory>src/main/docker</docker.directory>
```
Comment (joaquincasares) [inline src/server/pom.xml]: This looks good if the `docker.directory` will be changed frequently. What was the use case you had in mind? Or was it just due to the reorg?

File: src/server/src/main/docker/Dockerfile [hunk 5]
```diff
@@ -35,21 +37,35 @@ ENV REAPER_SEGMENT_COUNT=200 \
...
     REAPER_DB_PASSWORD="" \
+    REAPER_METRICS_ENABLED=false \
+    REAPER_METRICS_FREQUENCY="1 minute" \
+    REAPER_METRICS_REPORTERS="[]" \
     JAVA_OPTS=""
 
 ADD cassandra-reaper.yml /etc/cassandra-reaper.yml
 ADD entrypoint.sh /usr/local/bin/entrypoint.sh
-ADD append-persistence.sh /usr/local/bin/append-persistence.sh
+ADD configure-persistence.sh /usr/local/bin/configure-persistence.sh
```
Comment (joaquincasares) [inline src/server/src/main/docker/Dockerfile:52]: I like `configure` a lot better. Nice job fixing that.

File: src/server/src/main/docker/Dockerfile [hunk 6]
```diff
@@ -35,21 +37,35 @@ ENV REAPER_SEGMENT_COUNT=200 \
...
+ADD configure-metrics.sh /usr/local/bin/configure-metrics.sh
+
 RUN addgroup -S reaper && \
     adduser -S reaper reaper && \
     apk add --no-cache 'su-exec>=0.2' && \
     mkdir -p /var/lib/cassandra-reaper && \
-    chown reaper:reaper /etc/cassandra-reaper.yml /var/lib/cassandra-reaper /usr/local/bin/entrypoint.sh /usr/local/bin/append-persistence.sh && \
-    chmod u+x /usr/local/bin/entrypoint.sh /usr/local/bin/append-persistence.sh
+    chown reaper:reaper \
```
Comment (joaquincasares) [inline src/server/src/main/docker/Dockerfile:59]: Ah, good catch.

File: src/server/src/main/docker/configure-persistence.sh [hunk 7]
```diff
@@ -3,20 +3,29 @@
...
+activateQueryLogger: ${REAPER_ACTIVATE_QUERY_LOGGER}
+
 cassandra:
   clusterName: ${REAPER_CASS_CLUSTER_NAME}
   contactPoints: ${REAPER_CASS_CONTACT_POINTS}
   keyspace: ${REAPER_CASS_KEYSPACE}
 EOT
 
-if [ "true" = "${REAPER_CASS_AUTH_ENABLED}" ]; then
+    if [ "true" = "${REAPER_CASS_AUTH_ENABLED}" ]; then
```
Comment (joaquincasares) [inline src/server/src/main/docker/configure-persistence.sh]: The new file looks a bit off with the indentations. Perhaps a series of comment lines like `# BEGIN... # END`?

File: src/server/src/main/docker/entrypoint.sh [hunk 8]
```diff
@@ -2,9 +2,9 @@
 
 if [ "$1" = 'cassandra-reaper' ]; then
     set -x
-    su-exec reaper /usr/local/bin/append-persistence.sh
-    exec su-exec reaper java -jar ${JAVA_OPTS} \
-        /usr/local/lib/cassandra-reaper.jar server /etc/cassandra-reaper.yml
+    su-exec reaper /usr/local/bin/configure-persistence.sh
```
Comment (joaquincasares) [inline src/server/src/main/docker/entrypoint.sh:5]: Any idea on why some are `su-exec` and one is `exec su-exec`? Does the former fork the process or something?

File: src/server/src/main/docker/entrypoint.sh [hunk 9]
```diff
@@ -2,9 +2,9 @@
...
+    su-exec reaper /usr/local/bin/configure-metrics.sh
+    exec su-exec reaper java ${JAVA_OPTS} -jar /usr/local/lib/cassandra-reaper.jar server /etc/cassandra-reaper.yml
```
Comment (joaquincasares) [inline src/server/src/main/docker/entrypoint.sh]: Mick mentioned wanting to use the `bin/start` file. Might want to be updated here.

Also a `\ ` newline may be good to highlight the different arguments.

File: src/docs/content/docs.md [hunk 10]
```diff
@@ -352,41 +350,103 @@ Source code for all the REST resources can be found from package com.spotify.rea
...
-First, build the Docker image and add it to your local image cache using the
-`cassandra-reaper:latest` tag:
+Building the Reaper Docker Image requires that the Reaper JAR be present in the _src/server/target_ directory. When Reaper is built either by Maven or the [Docker build image](building-reaper-packages-with-docker), the JAR file will be placed in the _target_ directory as part of the build process.
 
-```mvn clean package docker:build```
+To build the Reaper Docker Image which is then added to the local image cache using the `cassandra-reaper:latest` tag, run the following commands from the top level directory:
+
+```bash
+cd src/server
+mvn docker:build
```
[the remark this answers is not in the document]
Comment (ossarga) [reply src/docs/content/docs.md]: Good point. Was originally thinking to use the docker image to build the JARs as it takes care of the need to install the third packages required to build Reaper. Will reword this as suggested.

File: src/docs/content/docs.md [hunk 11]
```diff
@@ -397,19 +457,27 @@ http://127.0.0.1:8080/webui/
 
 When adding the Cassandra node to the Reaper UI, use the IP address found via:
 
-```docker-compose run nodetool status```
+```bash
```
[the remark this answers is not in the document]
Comment (ossarga) [reply src/docs/content/docs.md]: No worries.

File: src/packaging/docker-build/docker-compose.yml [hunk 12]
```diff
@@ -1,4 +1,4 @@
-version: '2.1'
+version: '2.2'
```
[the remark this answers is not in the document]
Comment (ossarga) [reply src/packaging/docker-build/docker-compose.yml]: The `reaper-build-packages` service references the _src/packaging/docker-build/docker-compose.yml_ which is set to version 2.2. Is this change ok or should we drop the referenced Docker Compose file down to 2.1?

File: src/packaging/docker-services/generate-ssl-stores/docker-entrypoint.sh [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+CASSANDRA_KEYSTORE=${WORKDIR}/ssl-stores/cassandra-server-keystore.jks
+REAPER_KEYSTORE=${WORKDIR}/ssl-stores/reaper-server-keystore.jks
+GENERIC_TRUSTSTORE=${WORKDIR}/ssl-stores/generic-server-truststore.jks
+
+CA_CERT_CONFIG=${WORKDIR}/ca_cert.conf
+ROOT_CA_CERT=${WORKDIR}/ssl-stores/ca-cert
+
+echo "Generic Certificate Authority configuration"
+cat ${CA_CERT_CONFIG}
```
[the remark this answers is not in the document]
Comment (ossarga) [reply src/packaging/docker-services/generate-ssl-stores/docker-entrypoint.sh]: Thanks

File: src/docs/content/docs.md [hunk 14]
```diff
@@ -352,41 +350,103 @@ Source code for all the REST resources can be found from package com.spotify.rea
...
+Start the Cassandra cluster which encrypts both the JMX and Native Protocol:
+
+```bash
+docker-compose up cassandra-ssl
+```
+
+The `nodetool-ssl` Docker Compose service can be used to check on the Cassandra node's status:
+
+```bash
+docker-compose run nodetool-ssl status
```
[the remark this answers is not in the document]
Comment (ossarga) [reply src/docs/content/docs.md:446]: Thanks

File: src/server/pom.xml (hunk 4 above)
Comment (ossarga) [reply src/server/pom.xml]: This was kind of part of the reorg. It allows the following command to be run from the top-level of the repository.
```bash
mvn -f src/server/pom.xml clean package docker:build -Ddocker.directory=src/server/src/main/docker
```

File: src/server/src/main/docker/Dockerfile (hunk 5 above)
Comment (ossarga) [reply src/server/src/main/docker/Dockerfile:52]: No worries. Thanks.

File: src/server/src/main/docker/Dockerfile (hunk 6 above)
Comment (ossarga) [reply src/server/src/main/docker/Dockerfile:59]: Cheers.

File: src/packaging/docker-services/generate-ssl-stores/docker-entrypoint.sh [hunk 15]
```diff
@@ -0,0 +1,59 @@
+#!/bin/sh
```
[the remark this answers is not in the document]
Comment (ossarga) [reply src/packaging/docker-services/generate-ssl-stores/docker-entrypoint.sh:1]: Thanks!

File: src/packaging/docker-services/generate-ssl-stores/docker-entrypoint.sh [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+
+set -ex
+
```
[the remark this answers is not in the document]
Comment (ossarga) [reply src/packaging/docker-services/generate-ssl-stores/docker-entrypoint.sh:4]: Good idea. Have added comments at the top of file to explain those points mentioned.

File: src/packaging/docker-services/initialize-reaper_db-ssl/initialize-reaper_db-ssl.env [hunk 17]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+
+CASSANDRA_KEYSTORE_ALIAS=cassandra
+CASSANDRA_KEYSTORE_PASSWORD=keypassword
+
+# use the same DC value set in the cassandra-ssl.env file
+CASSANDRA_DC=docker-compose-1
```
[the remark this answers is not in the document]
Comment (ossarga) [reply src/packaging/docker-services/initialize-reaper_db-ssl/initialize-reaper_db-ssl.env]: Good point. I will change the naming of this to `datacenter1`.

File: src/packaging/resource/ca_cert.conf (hunk 3 above)
Comment (ossarga) [reply src/packaging/resource/ca_cert.conf]: 👍

File: src/server/src/main/docker/configure-persistence.sh (hunk 7 above)
Comment (ossarga) [reply src/server/src/main/docker/configure-persistence.sh]: I was on the fence with the indentations as it was. I have added in the comments as suggested.

File: src/server/src/main/docker/entrypoint.sh (hunk 8 above)
Comment (ossarga) [reply src/server/src/main/docker/entrypoint.sh:5]: Possibly. Not 100% sure to be honest. Definitely the later replaces the current shell with the command.

File: src/server/src/main/docker/entrypoint.sh (hunk 9 above)
Comment (ossarga) [reply src/server/src/main/docker/entrypoint.sh]: If it is ok, can we leave off the change use the `bin/start` file here please. Reason being is that file will need some changes to allow addition `JAVA_OPTS` to be passed to it.  We should raise a separate PR to make that change.

In the meantime, the suggestion to add a `\` newline to highlight the different arguments has been applied.

File: src/packaging/docker-compose.yml [hunk 18]
```diff
@@ -15,32 +15,97 @@ services:
       - "9042:9042"
     volumes:
       - ./data/cassandra:/var/lib/cassandra
-      - ./docker/cassandra/jmxremote.access:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/management/jmxremote.access
-      - ./docker/cassandra/jmxremote.password:/etc/cassandra/jmxremote.password
+      - ./docker-services/cassandra/jmxremote.access:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/management/jmxremote.access
+      - ./docker-services/cassandra/jmxremote.password:/etc/cassandra/jmxremote.password
+
+  cassandra-ssl:
```
[the remark this answers is not in the document]
Comment (ossarga) [reply src/packaging/docker-compose.yml]: Good point. Have created a _common-services.yml_ file which contains the services to that are extended by _docker-compose.yml_.

File: src/packaging/docker-services/cassandra-ssl/Dockerfile [hunk 19]
```diff
@@ -0,0 +1,20 @@
+FROM cassandra:3.11
```
[the remark this answers is not in the document]
Comment (ossarga) [reply src/packaging/docker-services/cassandra-ssl/Dockerfile]: Fair call. Have removed this file and extended the common Cassandra service.

File: src/packaging/docker-services/cassandra-ssl/cassandra-ssl.env [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+#CASSANDRA_LISTEN_ADDRESS=cassandra
+#CASSANDRA_SEEDS=cassandra
+
+# leave blank to auto configure
+
+CASSANDRA_BROADCAST_ADDRESS
+CASSANDRA_RPC_ADDRESS
+
+# name of this Cassandra cluster
+CASSANDRA_CLUSTER_NAME='cassandra-ssl-cluster'
```
[the remark this answers is not in the document]
Comment (ossarga) [reply src/packaging/docker-services/cassandra-ssl/cassandra-ssl.env]: No worries. Fixed! Thanks.

File: src/packaging/docker-services/initialize-reaper_db-ssl/docker-entrypoint.sh [hunk 21]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (ossarga) [reply src/packaging/docker-services/initialize-reaper_db-ssl/docker-entrypoint.sh]: Done. Made the SSL components switchable in the _docker-entry.sh_ file using an environment var.

File: src/packaging/docker-services/nodetool-ssl/Dockerfile (hunk 1 above)
Comment (ossarga) [reply src/packaging/docker-services/nodetool-ssl/Dockerfile]: Done.

File: src/packaging/docker-services/reaper-ssl/reaper-ssl.env (hunk 2 above)
Comment (ossarga) [reply src/packaging/docker-services/reaper-ssl/reaper-ssl.env]: Ok. Done.

File: src/packaging/docker-build/docker-compose.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (joaquincasares) [reply src/packaging/docker-build/docker-compose.yml]: I would have downgraded `src/packaging/docker-build/docker-compose.yml` to 2.1 and seen if that worked. But since there's another file that's cross-referenced, I'd assume it's already at its lowest version.

This comment was more cosmetic, but always good to try to keep the version dependency lower when/if possible.

File: src/packaging/docker-services/generate-ssl-stores/docker-entrypoint.sh [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+GENERIC_TRUSTSTORE=${WORKDIR}/ssl-stores/generic-server-truststore.jks
+
+CA_CERT_CONFIG=${WORKDIR}/ca_cert.conf
+ROOT_CA_CERT=${WORKDIR}/ssl-stores/ca-cert
+
+echo "Generic Certificate Authority configuration"
+cat ${CA_CERT_CONFIG}
+echo
+
+rm -rf ${WORKDIR}/ssl-stores/*
```
[the remark this answers is not in the document]
Comment (joaquincasares) [reply src/packaging/docker-services/generate-ssl-stores/docker-entrypoint.sh]: +1  on having that prompt and flow. Defaulting to [Y/n] on Yes to delete would be most ideal.

File: src/server/pom.xml (hunk 4 above)
Comment (joaquincasares) [reply src/server/pom.xml]: Ah, thanks for clarifying that for me.

Comment (ossarga) [reply src/server/pom.xml]: No worries.

File: src/packaging/docker-build/docker-compose.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (ossarga) [reply src/packaging/docker-build/docker-compose.yml]: That is a fair call. I will downgrade _src/packaging/docker-build/docker-compose.yml_ and _src/packaging/docker-build/docker-compose.yml_ to version 2.1, then test to make sure it still works.

File: src/packaging/docker-services/generate-ssl-stores/docker-entrypoint.sh (hunk 22 above)
[the remark this answers is not in the document]
Comment (ossarga) [reply src/packaging/docker-services/generate-ssl-stores/docker-entrypoint.sh]: Instead of clearing out the contents of the directory, I will check if any of the files can be left and overridden or if they all need to be deleted. That way we delete only what is necessary instead of the entire directory.

File: src/server/src/main/docker/Dockerfile [hunk 23]
```diff
@@ -13,6 +13,8 @@ ENV REAPER_SEGMENT_COUNT=200 \
     REAPER_INCREMENTAL_REPAIR=false \
     REAPER_DATACENTER_AVAILABILITY=ALL \
     REAPER_ENABLE_DYNAMIC_SEEDS=true \
+    REAPER_REPAIR_MANAGER_SCHEDULEING_INTERVAL_SECONDS=30 \
```
Comment (joaquincasares) [inline src/server/src/main/docker/Dockerfile]: Typo on `scheduleing`.

File: src/server/src/main/docker/cassandra-reaper.yml [hunk 24]
```diff
@@ -9,6 +9,7 @@ storageType: ${REAPER_STORAGE_TYPE}
 enableCrossOrigin: ${REAPER_ENABLE_CORS}
 incrementalRepair: ${REAPER_INCREMENTAL_REPAIR}
 enableDynamicSeedList: ${REAPER_ENABLE_DYNAMIC_SEEDS}
+repairManagerSchedulingIntervalSeconds: ${REAPER_REPAIR_MANAGER_SCHEDULEING_INTERVAL_SECONDS}
```
Comment (joaquincasares) [inline src/server/src/main/docker/cassandra-reaper.yml]: Typo here as well.

File: src/packaging/docker-compose.yml [hunk 25]
```diff
@@ -2,50 +2,92 @@ version: '2.1'
...
-      - "8081:8081"
 
   reaper-build-packages:
     extends:
       file: ./docker-build/docker-compose.yml
-      service: build
\ No newline at end of file
+      service: build
+
+  reaper-ssl:
```
[the remark this answers is not in the document]
Comment (ossarga) [reply src/packaging/docker-compose.yml]: Sure thing.

File: src/packaging/docker-services/generate-ssl-stores/docker-entrypoint.sh (hunk 16 above)
[the remark this answers is not in the document]
Comment (ossarga) [reply src/packaging/docker-services/generate-ssl-stores/docker-entrypoint.sh:4]: Awesome. Great to hear it is useful!

File: src/packaging/docker-services/initialize-reaper_db/docker-entrypoint.sh [hunk 26]
```diff
@@ -1,11 +1,55 @@
 #!/usr/bin/env bash
 
-set -x
+set -xe
```
[the remark this answers is not in the document]
Comment (ossarga) [reply src/packaging/docker-services/initialize-reaper_db/docker-entrypoint.sh]: Thanks :)

File: src/packaging/docker-services/initialize-reaper_db/docker-entrypoint.sh [hunk 27]
```diff
@@ -1,11 +1,55 @@
...
+        -srcstorepass ${CASSANDRA_KEYSTORE_PASSWORD} \
+        -destkeystore ${PKCS12_KEYSTORE} -deststoretype PKCS12 -destalias ${PKCS12_KEYSTORE_USERNAME} \
+        -deststorepass ${PKCS12_KEYSTORE_PASSWORD}
+
+    # extract key and cert from the PKCS and place them in heir own files
+    openssl pkcs12 -in ${PKCS12_KEYSTORE} -nokeys -out ${CQL_CERT_PEM} -passin pass:${PKCS12_KEYSTORE_PASSWORD}
+    openssl pkcs12 -in ${PKCS12_KEYSTORE} -nocerts -nodes -out ${CQL_KEY_PEM} -passin pass:${PKCS12_KEYSTORE_PASSWORD}
+
+    # make cqlsh expect the generated key and cert files
+cat <<EOT >> ${CQLSH_RC}
```
[the remark this answers is not in the document]
Comment (ossarga) [reply src/packaging/docker-services/initialize-reaper_db/docker-entrypoint.sh]: Sure. Done.

File: src/server/src/main/docker/Dockerfile (hunk 23 above)
Comment (ossarga) [reply src/server/src/main/docker/Dockerfile]: Hah! Well spotted thanks!

File: src/server/src/main/docker/cassandra-reaper.yml (hunk 24 above)
Comment (ossarga) [reply src/server/src/main/docker/cassandra-reaper.yml]: Nicely spotted again!

File: src/packaging/docker-compose.yml [hunk 28]
```diff
@@ -2,50 +2,92 @@ version: '2.1'
...
+      - ./docker-services/cassandra-ssl/cassandra-ssl.env
     volumes:
-      - ./data/cassandra:/var/lib/cassandra
-      - ./docker-services/cassandra/jmxremote.access:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/management/jmxremote.access
-      - ./docker-services/cassandra/jmxremote.password:/etc/cassandra/jmxremote.password
+      - ./ssl-stores/cassandra-server-keystore.jks:/etc/ssl/cassandra-server-keystore.jks
+      - ./ssl-stores/generic-server-truststore.jks:/etc/ssl/generic-server-truststore.jks
+      - ./docker-services/cassandra-ssl/docker-entrypoint.sh:/docker-alt-entrypoint.sh
 
   cqlsh:
```
[the remark this answers is not in the document]
Comment (ossarga) [reply src/packaging/docker-compose.yml:21]: True. Added one in.

What I have done is added in a `cqlsh-common` service into _common-services.yml_. I then derive the `cqlsh`, `cqlsh-ssl`, `cqlsh-initialize-reaper_db`, and `cqlsh-initialize-reaper_db-ssl` services from that. Note that the `cqlsh-initialize-reaper_db`, and `cqlsh-initialize-reaper_db-ssl` replace the `initialize-reaper_db` and `initialize-reaper_db-ssl` services respectively.

The reason I did this was to avoid duplicating the CQLSH SSL connection process. The changes are in https://github.com/thelastpickle/cassandra-reaper/pull/158/commits/ec9d1e83239d4f155d5f568cccf3af18863489f7

@joaquincasares - let me know what you think.

[the remark this answers is not in the document]
Comment (joaquincasares) [reply src/packaging/docker-compose.yml:21]: Oh, that was a clever reuse. Looks good! :)

**Comment by ossarga** [comment]
Merging this PR as per feedback from @joaquincasares
